file-attribute-caching/src/main/kotlin/com/pkware/file/fileAttributeCaching/FileAttributeCachingFileSystemProvider.kt

```diff
@@ -0,0 +1,465 @@
+
+        // If we have both target and source caching paths we copy the attributes from source to target and run
+        // Files.move(source, target, *newOptions).
+        if (options.contains(StandardCopyOption.COPY_ATTRIBUTES) &&
+            source is FileAttributeCachingPath &&
+            target is FileAttributeCachingPath
+        ) {
+            // Filter out StandardCopyOption.COPY_ATTRIBUTES here because we dont want the moved file to repopulate
+            // the cache from the delegate provider/filesystem.
+            val newOptions = options.filter {
+                it != StandardCopyOption.COPY_ATTRIBUTES
+            }.toTypedArray()
+
+            source.copyCachedAttributesTo(target)
+            Files.move(actualSourcePath, actualTargetPath, *newOptions)
+        } else {
+            // If the StandardCopyOption.COPY_ATTRIBUTES option is not selected, there is no need to cache the
+            // attributes for the moved file.
+            Files.move(actualSourcePath, actualTargetPath, *options)
+        }
+    }
+
+    override fun isSameFile(path: Path, path2: Path): Boolean {
+        val actualPath = if (path is FileAttributeCachingPath) path.delegate else path
+        val actualPath2 = if (path2 is FileAttributeCachingPath) path2.delegate else path2
+
+        return Files.isSameFile(actualPath, actualPath2)
+    }
+
+    /**
+     * Tells whether a file is considered to be hidden. The exact definition of hidden is platform or provider
+     * dependent.
+     *
+     * On Windows a file is considered hidden if it isn't a directory and the DOS hidden attribute is set.
+     *
+     * On UNIX a file is considered to be hidden if its name begins with a period character ('.').
+     *
+     * Depending on the implementation, this method may access the [path] to determine if the file is
+     * considered hidden.
+     *
+     * @param path The path to check.
+     * @return `true` if the path is hidden, `false` otherwise.
+     * @throws IOException if an error occurs while accessing the underlying delegate provider or the provided path was
+     * not a [FileAttributeCachingPath].
+     */
+    @Throws(IOException::class)
+    override fun isHidden(path: Path): Boolean = if (path is FileAttributeCachingPath) {
+        if (isWindowsOS) {
+            val attributesMap = path.getAllAttributesMatchingName("dos:*") ?: throw IOException(
+                "Could not get dos attributes from delegate filesystem."
+            )
+            val isHidden = attributesMap["dos:hidden"] as Boolean && !(attributesMap["directory"] as Boolean)
+            isHidden
+        } else {
+            val delegatePath = path.delegate
+            val delegateProvider = delegatePath.fileSystem.provider()
+            delegateProvider.isHidden(delegatePath)
+        }
+    } else {
+        throw IOException("Path was not a FileAttributeCachingPath, could not read hidden status.")
```

Comment:
```suggestion
        throw UnsupportedOperationException("Path was not a FileAttributeCachingPath, could not read hidden status.")
```

Reply:
Save the use of IOExceptions for problems with I/O

Reply:
This function doesn't throw UnsupportedOperationException in the base class

Reply:
read and set attributes do, so I dont really want to change this

Reply:
I will add clarification to this message though like the other feedback that was suggested

Reply:
You are accounting for behavior not specified in the base class. If you're _adding_ behavior to your implementation (namely, adding a `RuntimeException` which doesn't modify the signature), it's perfectly fine to deviate.

Reply:
Also, I think your earlier throw of `IOException` is fine in this method, as it does represent a legitimate IO problem.

Reply:
And idk, maybe even this is preferred?

```suggestion
                FileSystems.getDefault().provider().isHidden(path)
```